src/dune_file_watcher/dune_file_watcher.ml

```diff
@@ -575,6 +587,88 @@ let create_fsevents ?(latency = 0.2) ~(scheduler : Scheduler.t) () =
   ; sync_table
   }
 
+let winwatch_patterns =
+  List.rev_map ~f:Re.Str.regexp_case_fold
+    [ {|^_opam|}
+    ; {|/_opam|}
+    ; {|^_esy|}
+    ; {|/_esy|}
+    ; {|^\.#.*|} (* Such files can be created by Emacs and also Dune itself. *)
+    ; {|/\.#.*|}
+    ; {|~$|}
+    ; {|^#[^#]*#$|}
+    ; {|/#[^#]*#$|}
+    ; {|^4913$|} (* https://github.com/neovim/neovim/issues/3460 *)
+    ; {|/4913$|}
+    ; {|^.git|}
+    ; {|/.git|}
+    ; {|^.hg|}
+    ; {|/.hg|}
+    ; {|^c:/windows|}
+    ]
+
+let winwatch_filter path =
+  let path =
+    String.concat ~sep:"/"
+      (String.split_on_char ~sep:'\\'
+         (String.lowercase_ascii (Path.to_string path)))
+  in
+  List.for_all
+    ~f:(fun pattern ->
+      try
+        let _ = Re.Str.search_forward pattern path 0 in
+        false
+      with Not_found -> true)
+    winwatch_patterns
+
+let winwatch_callback ~(scheduler : Scheduler.t) ~sync_table ~dir action
+    filename =
+  let filename = Filename.concat dir filename in
+  let localized_path =
+    Path.Expert.try_localize_external (Path.of_string filename)
+  in
+  match localized_path with
+  | In_build_dir _ -> (
+    if Fs_sync.is_special_file_fsevents localized_path then
+      match action with
+      | Winwatch.Event.Added | Modified -> (
+        match Fs_sync.consume_event sync_table filename with
+        | None -> ()
+        | Some id ->
+          scheduler.thread_safe_send_emit_events_job (fun () -> [ Sync id ]))
+      | Removed | Renamed_new | Renamed_old -> ())
+  | path ->
+    if winwatch_filter path && not (Path.is_directory path) then
+      scheduler.thread_safe_send_emit_events_job (fun () ->
+          let kind =
+            match action with
+            | Winwatch.Event.Added | Renamed_new -> Fs_memo_event.Created
+            | Removed | Renamed_old -> Deleted
+            | Modified -> File_changed
+          in
+          [ Fs_memo_event { kind; path } ])
+
+let create_winwatch ~(scheduler : Scheduler.t) =
+  let sync_table = Table.create (module String) 64 in
+  let iocp = Winwatch.Iocp.create () in
+  let source =
+    let dir = Path.to_absolute_filename Path.root in
+    match
+      Winwatch.create dir ~f:(winwatch_callback ~scheduler ~sync_table ~dir)
+    with
+    | None -> assert false
+    | Some t -> t
+  in
+  Winwatch.start source iocp;
+  scheduler.spawn_thread (fun () ->
+      match Winwatch.Iocp.run iocp with
+      | Ok () -> ()
+      | Error _ -> (* Winwatch.Iocp.run never returns *) assert false);
```

Comment:
What are exactly the semantics of `run`? It looks like we could make it return `_`, or some kind of empty type.